File: src/main/java/com/github/dkoval/leetcode/challenge/UncrossedLines.java
package com.github.dkoval.leetcode.challenge;

/**
 * <a href="https://leetcode.com/problems/uncrossed-lines/">Uncrossed Lines</a>
 * <p>
 * You are given two integer arrays nums1 and nums2. We write the integers of nums1 and nums2 (in the order they are given) on two separate horizontal lines.
 * <p>
 * We may draw connecting lines: a straight line connecting two numbers nums1[i] and nums2[j] such that:
 * <ul>
 *  <li>nums1[i] == nums2[j], and</li>
 *  <li>the line we draw does not intersect any other connecting (non-horizontal) line.</li>
 *  </ul>
 * Note that a connecting line cannot intersect even at the endpoints (i.e., each number can only belong to one connecting line).
 * <p>
 * Return the maximum number of connecting lines we can draw in this way.
 * <p>
 * Constraints:
 * <ul>
 *  <li>1 <= nums1.length, nums2.length <= 500</li>
 *  <li>1 <= nums1[i], nums2[j] <= 2000</li>
 * </ul>
 */
public interface UncrossedLines {

    int maxUncrossedLines(int[] nums1, int[] nums2);

    class UncrossedLinesDPTopDown implements UncrossedLines {

        @Override
        public int maxUncrossedLines(int[] nums1, int[] nums2) {
            int n1 = nums1.length;
            int n2 = nums2.length;

            // DP: top-down
            Integer[][] dp = new Integer[n1][n2];
            return calculate(nums1, nums2, 0, 0, dp);
        }

        private int calculate(int[] nums1, int[] nums2, int idx1, int idx2, Integer[][] dp) {
            // base case
            if (idx1 >= nums1.length || idx2 >= nums2.length) {
                return 0;
            }

            // already solved?
            if (dp[idx1][idx2] != null) {
                return dp[idx1][idx2];
            }

            int best = 0;
            if (nums1[idx1] == nums2[idx2]) {
                best = 1 + calculate(nums1, nums2, idx1 + 1, idx2 + 1, dp);
            } else {
                // option #1: leave nums1[idx1] and try to find a match in nums2[idx2 : ]
                best = Math.max(best, calculate(nums1, nums2, idx1, idx2 + 1, dp));
                // option #2: leave nums2[idx2] and try to find a match in nums1[idx1 : ]
                best = Math.max(best, calculate(nums1, nums2, idx1 + 1, idx2, dp));
            }
            return dp[idx1][idx2] = best;
        }
    }
}

File: src/main/kotlin/com/github/dkoval/leetcode/challenge/UncrossedLines.kt
package com.github.dkoval.leetcode.challenge

import kotlin.math.max

object UncrossedLinesDPBottomUp : UncrossedLines {

    // Resource: https://www.youtube.com/watch?v=jLv-5coG-qQ
    override fun maxUncrossedLines(nums1: IntArray, nums2: IntArray): Int {
        val n1 = nums1.size
        val n2 = nums2.size

        // dp[i][j] denotes the maximum number of uncrossed connecting lines between A[0:i] and B[0:j]
        val dp = Array(n1 + 1) { IntArray(n2 + 1) { 0 } }
        for (i in 1..n1) {
            for (j in 1..n2) {
                dp[i][j] =
                    if (nums1[i - 1] == nums2[j - 1]) 1 + dp[i - 1][j - 1]
                    else max(dp[i][j - 1], dp[i - 1][j])
            }
        }
        return dp[n1][n2] // the largest element in dp
    }
}